R/toast.R

```diff
@@ -0,0 +1,462 @@
+#'   object provided that the `id` was set when created/shown.
+#' @param ... Reserved for future extensions (currently ignored).
+#' @param session Shiny session object.
+#'
+#' @return `show_toast()` Invisibly returns the toast ID (string) that can be
+#'   used with `hide_toast()`.
+#'
+#' @family Toast components
+#' @describeIn show_toast Show a toast notification.
+#' @export
+show_toast <- function(
+  toast,
+  ...,
+  session = shiny::getDefaultReactiveDomain()
+) {
+  rlang::check_dots_empty()
+
+  if (!inherits(toast, "bslib_toast")) {
+    toast <- toast(toast)
+  }
+
+  if (length(toast$body) == 0 && length(toast$header) == 0) {
+    rlang::warn("`toast` has no content; no toast to show.")
+    return(invisible(NULL))
+  }
+
+  toast$id <- toast$id %||% toast_random_id()
+
+  toasted <- processDeps(toast, session)
+
+  options <- dropNulls(list(autohide = toast$autohide, delay = toast$duration))
+
+  data <- list(
+    html = toasted$html,
+    deps = toasted$deps,
+    options = options,
+    position = toast$position,
+    id = toast$id
+  )
+
+  # Show toast immediately
+  session$sendCustomMessage("bslib.show-toast", data)
+  invisible(toast$id)
+}
+
+#' @describeIn show_toast Hide a toast notification by ID.
+#' @export
+hide_toast <- function(id, ..., session = shiny::getDefaultReactiveDomain()) {
+  rlang::check_dots_empty()
+
+  if (inherits(id, "bslib_toast")) {
+    if (is.null(id$id)) {
+      rlang::abort("Cannot hide a toast without an ID. Provide the toast ID.")
+    }
+    id <- id$id
+  }
+  if (is.null(id)) {
+    rlang::warn("`id` is NULL; no toast to hide.")
+    return(invisible(NULL))
+  }
+  session$sendCustomMessage("bslib.hide-toast", list(id = id))
+  invisible(id)
+}
+
+toast_component_header <- function(x) {
+  # Status text (small muted text)
+  status_text <- if (!is.null(x$status)) {
+    tags$small(class = "text-muted text-end", x$status)
+  }
+
+  tagList(
+    if (!is.null(x$icon)) span(class = "toast-header-icon", x$icon),
+    strong(
+      class = "me-auto",
+      class = if (!is.null(x$icon)) "ms-2",
+      x$title
+    ),
+    status_text
+  )
+}
+
+# Internal function to build toast HTML structure
+toast_component <- function(
+  body,
+  header = NULL,
+  icon = NULL,
+  type = NULL,
+  closable = TRUE,
+  id = NULL,
+  class = NULL,
+  attribs = list()
+) {
+  # Set ARIA attributes based on toast urgency level
+  # - danger toasts use role="alert" + aria-live="assertive" to immediately
+  #   interrupt screen readers (critical errors requiring immediate attention)
+  # - other toasts use role="status" + aria-live="polite" to announce updates
+  #   without interrupting (non-critical, announced when convenient)
+  aria_role <- if (!is.null(type) && type == "danger") "alert" else "status"
+  aria_live <- if (!is.null(type) && type == "danger") "assertive" else "polite"
+
+  type_class <- if (!is.null(type)) {
+    paste0("text-bg-", type)
+  }
```

Comment:
If I'm being honest, I'm not totally in love with the "loud" body-bg color this produces, but I'm also not sure it's worth doing anything "smarter"?

<img width="367" height="97" alt="Screenshot 2025-10-31 at 3 48 28 PM" src="https://github.com/user-attachments/assets/0d9aaf7f-9713-46d7-a6bc-535fdfb0d4a2" />

Reply:
Yeah, I would also say that this isn't my favorite appearance, but importantly this is what we get out-of-the-box from Bootstrap when you [follow their recommendation](https://getbootstrap.com/docs/5.3/components/toasts/#color-schemes) and add `.text-bg-success` to the `.toast` container itself.

My take is that this is a simple and easy mechanism and any further work we do to customize the appearance would not only be more work to maintain over time, but would also be potentially fragile across Bootstrap themes and versions (e.g. Bootswatch, etc.).